src/Symfony/Component/AssetMapper/Compiler/JavaScriptImportPathCompiler.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\AssetMapper\Compiler;
+
+use Symfony\Component\AssetMapper\AssetMapper;
+use Symfony\Component\AssetMapper\MappedAsset;
+
+/**
+ * Resolves import paths in JS files.
+ *
+ * @experimental
+ *
+ * @author Ryan Weaver <ryan@symfonycasts.com>
+ */
+class JavaScriptImportPathCompiler implements AssetCompilerInterface
+{
+    use AssetCompilerPathResolverTrait;
+
+    // https://regex101.com/r/VFdR4H/1
+    private const IMPORT_PATTERN = '/(?:import\s+(?:(?:\*\s+as\s+\w+|[\w\s{},*]+)\s+from\s+)?|\bimport\()\s*[\'"`](\.\/[^\'"`]+|(\.\.\/)+[^\'"`]+)[\'"`]\s*[;\)]?/m';
+
+    public function compile(string $content, MappedAsset $asset, AssetMapper $assetMapper): string
+    {
+        return preg_replace_callback(self::IMPORT_PATTERN, function ($matches) use ($asset, $assetMapper) {
+            $resolvedPath = $this->resolvePath(\dirname($asset->getLogicalPath()), $matches[1]);
+
+            $finalResolvedPath = $resolvedPath;
+            if (!str_ends_with($finalResolvedPath, '.js')) {
```

Comment:
Native imports don't allow omitting the `.js` extension. I would say that this addition of a missing extension goes against the spirit of making you write JS in the modern way (plus, it is broken if you file uses the `.mjs` extension for instance)

Reply:
Sigh. Yea... we need to find the right balance here. I'm trying to make coding with importmaps identical to coding with Webpack. But... bending standards/rules is shaky business. But it's painful to have my editor autocomplete `import './foo'` and my editor is even aware what that means... then it would simply break when I refresh the page with a 404 to `/assets/foo`. @dunglas?

> plus, it is broken if you file uses the .mjs extension for instance

This could be fixed by first looking to see if the file exists without changing it. Then checking for `.js` and `.mjs`.

Reply:
I would consider removing this functionality (adding the `.js` on behalf of the user) if we added a "strict" type (enabled by default) to asset mapper where if we don't recognize a path, we throw an error. Then, instead of things just "not working" + getting a 404 for `/asset/foo` when you refresh, we could give a hard error:

> The import for `./foo` in `assets/app.js` could not be found. Did you mean `./foo.js`?

Reply:
I agree with @stof. The goal here is not to replace Webpack but to "use the platform" and promote using standard modern JS as much as possible. Standard editors and browser developers decided to not support this feature, so we shouldn't.
This philosophy will also allows
 to keep this component as small and as simple as possible.

To be honest I would even go as far as **not** supporting the `./localfile.js` syntax, as Rails.

Reply:
Ok, I will remove this - and add a strict mode.

> To be honest I would even go as far as not supporting the ./localfile.js syntax, as Rails.

Using `import './localfile.js` IS valid JS though, no? Like, if we were building everything directly inside of `public/` (with no moving around of files), using `./localfile.js` would work out-of-the-box. The only reason it doesn't work out-of-the-box for us is because we're digesting the files. And *making* this work is quite easy: we "notice" the imported file and make sure it's in the importmap.

Reply:
Sure but the developper isn't working in the public directory and I'm not fond of editing the JS files on the fly. Import Maps allow to have something simpler. But I'll not block this feature!

Reply:
> Sure but the developper isn't working in the public directory and I'm not fond of editing the JS files on the fly. Import Maps allow to have something simpler. But I'll not block this feature!

If we remove the rewriting of `./foo` to `./foo.js` (which i AM about to remove), then there is no "editing JS files" I can happily say - and that's thanks to importmaps. If we see `import './foo.js'`, we won't change it. What we WILL do is add an importmap entry for that `foo.js` file automatically so that the browser can find the digested filename. So, I think we're both happy? :)

Reply:
Indeed, perfect!

Reply:
Note that for the part where your IDE autocomplete `./foo`, this is configurable (I know it is in PHPStorm, and I'm almost sure VS Code has an equivalent config option)